include/extractor/guidance/mergable_road_detector.hpp

```diff
@@ -0,0 +1,149 @@
+     */
+    bool RoadDataIsCompatible(const util::NodeBasedEdgeData &lhs_edge_data,
+                              const util::NodeBasedEdgeData &rhs_edge_data) const;
+
+    /* Detect traffic loops.
+     * Since OSRM cannot handle loop edges, we cannot directly see a connection between a node and
+     * itself. We need to skip at least a single node in between. */
+    bool IsTrafficLoop(const NodeID intersection_node, const MergableRoadData &road) const;
+
+    /* Detetor to check if we are looking at roads splitting up just prior to entering an
+     * intersection:
+     *
+     *        c
+     *      / |
+     * a -<   |
+     *      \ |
+     *        b
+     *
+     * A common scheme in OSRM is that roads spit up in separate ways when approaching an
+     * intersection. This detector tries to detect these narrow triangles which usually just offer a
+     * small island for pedestrians in the middle. */
+    bool IsNarrowTriangle(const NodeID intersection_node,
+                          const MergableRoadData &lhs,
+                          const MergableRoadData &rhs) const;
+
+    /* Detector to check for whether two roads are following the same direction.
+     * If roads don't end up right at a connected intersection, we could look at a situation like
+     *
+     *      __________________________
+     *     /
+     * ---
+     *     \__________________________
+     *
+     * This detector tries to find out about whether two roads are parallel after the separation */
+    bool HaveSameDirection(const NodeID intersection_node,
+                           const MergableRoadData &lhs,
+                           const MergableRoadData &rhs) const;
+
+    /* Detector for small traffic islands. If a road is splitting up, just to connect again later,
+     * we don't wan't to have this information within our list of intersections/possible turn
+     * locations.
+     *
+     *     ___________
+     * ---<___________>-----
+     *
+     *
+     * Would feel just like a single straight road to a driver and should be represented as such in
+     * our engine.*/
+    bool ConnectAgain(const NodeID intersection_node,
+                      const MergableRoadData &lhs,
+                      const MergableRoadData &rhs) const;
+
+    /* A negative detector, preventing a merge, trying to detect link roads between two main roads.
+     *
+     *  d - - - - - - - - e - f
+     *             . / '
+     *  a - - - b - - - - - - c
+     *
+     * The detector wants to prevent merges that are connected to `b-e` */
+    bool IsLinkRoad(const NodeID intersection_node, const MergableRoadData &road) const;
+
+    const util::NodeBasedDynamicGraph &node_based_graph;
+    const std::vector<QueryNode> &node_coordinates;
+    const IntersectionGenerator &intersection_generator;
+    const CoordinateExtractor &coordinate_extractor;
+
+    // name detection
+    const util::NameTable &name_table;
+    const SuffixTable &street_name_suffix_table;
```

Comment:
For all of these, if we only ever use them by ref in this header: you could forward declare them and only in the cpp impl. file include the headers. Not sure if it's worth it (especially with lto). Also maybe we should think of a `_fwd.hpp` header file all handler header could include, since most of the types used should be the same - what do you think?